modules/internal/String.chpl

```diff
@@ -17,572 +17,1436 @@
+      chpl_string_comm_get(dest, src_loc_id, src_addr, len.safeCast(size_t));
+      dest[len] = 0;
+      return dest;
   }
 
-  inline proc +(s: string, x: c_string) {
-    var cs = __primitive("string_concat", s.c_str(), x);
-    const ret = toString(cs);
-    // toString steals the c_string_copy, so no need to free it here.
-    return ret;
+  extern proc memmove(destination: c_ptr, const source: c_ptr, num: size_t);
+
+  // pointer arithmetic used for strings
+  inline proc +(a: c_ptr, b: integral) return __primitive("+", a, b);
+
+  config param debugStrings = false;
+}
+
+// Chapel Strings
+module String {
+  use BaseStringType;
+  use StringCasts;
+
+  inline proc chpl_debug_string_print(s: c_string) {
+    if debugStrings {
+      extern proc printf(format: c_string, x...);
+      printf("%s\n", s);
+    }
   }
 
-  inline proc +(s: string, x: numeric)
-    return s + x:string;
+  // TODO: We should be able to remove "ignore noinit", but doing so causes
+  // memory leaks in various places. Investigate why later and add noinit back
+  // in when possible.
+  pragma "string record"
+  pragma "ignore noinit"
+  pragma "no default functions"
+  record string {
+    var len: int = 0; // length of string in bytes
+    var _size: int = 0; // size of the buffer we own
+    var buff: bufferType = nil;
+    var owned: bool = true;
+
+    proc string(s: string, owned: bool = true) {
+      this.owned = owned;
+      const sRemote = s.locale.id != chpl_nodeID;
+      if !owned && sRemote {
+        // TODO: Should I just ignore the supplied value of owned instead of
+        //       halting?
+        halt("Must take ownership of a copy of a remote string");
+      }
+      var localBuff: bufferType = nil;
+      const sLen = s.len;
+      // Dont need to do anything if s is an empty string
+      if sLen != 0 {
+        localBuff = if sRemote
+          then copyRemoteBuffer(s.locale.id, s.buff, sLen)
+          else s.buff;
+      }
+      // We only need to copy when the buffer is local
+      // TODO: reinitString does more checks then we actually need but is still
+      // correct. Is it worth just doing the proper copy here?
+      this.reinitString(localBuff, sLen, sLen+1, needToCopy=sRemote);
+    }
 
-  inline proc +(x: numeric, s: string)
-    return x:string + s;
+    // This constructor can cause a leak if owned = false and needToCopy = true
+    proc string(buff: bufferType, length: int, size: int,
+                owned: bool = true, needToCopy: bool = true) {
+      this.owned = owned;
+      this.reinitString(buff, length, size, needToCopy);
+    }
 
-  inline proc +(s: string, x: enumerated)
-    return s + x:string;
+    proc ref ~string() {
+      if owned && !this.isEmptyString() {
+        on this {
+          chpl_mem_free(this.buff);
+        }
+      }
+    }
 
-  inline proc +(x: enumerated, s: string)
-    return x:string + s;
+    // This is assumed to be called from this.locale
```

Comment:
I'd like to see a comment about what this function should do for the different cases of this.owned and needToCopy.